File: ass_editor.js
#!/usr/bin/env node
// ass_editor.js
const fs = require('fs');
const path = require('path');

class ASSParser {
    constructor(filename) {
        this.filename = filename;
        this.header = [];
        this.events = [];
        this.rawLines = [];
    }

    parse() {
        const content = fs.readFileSync(this.filename, 'utf8');
        const lines = content.split(/\r?\n/);
        this.rawLines = lines;
        let inEvents = false;
        for (const line of lines) {
            if (line.startsWith('[Events]')) {
                inEvents = true;
                this.header.push(line);
                continue;
            }
            if (!inEvents) {
                this.header.push(line);
                continue;
            }
            if (line.startsWith('Format:')) {
                this.header.push(line);
                continue;
            }
            if (line.startsWith('Dialogue:') || line.startsWith('Comment:')) {
                const ev = this.parseEvent(line);
                if (ev) this.events.push(ev);
            } else {
                this.header.push(line);
            }
        }
        return this;
    }

    parseEvent(line) {
        const parts = line.split(',', 9);
        if (parts.length < 10) return null;
        const prefix = parts[0];
        const startStr = parts[1].trim();
        const endStr = parts[2].trim();
        const style = parts[3].trim();
        const text = parts.slice(9).join(',').trim(); // остаток после 9-й запятой
        const startMs = this.timeToMs(startStr);
        const endMs = this.timeToMs(endStr);
        return { startMs, endMs, style, text, prefix, raw: line };
    }

    timeToMs(ts) {
        if (!ts) return 0;
        const parts = ts.split(':');
        if (parts.length !== 3) return 0;
        const h = parseInt(parts[0]);
        const m = parseInt(parts[1]);
        const secPart = parts[2];
        let s = 0, ms = 0;
        if (secPart.includes('.')) {
            const [sec, millis] = secPart.split('.');
            s = parseInt(sec);
            ms = parseInt(millis.padEnd(3, '0').slice(0, 3));
        } else {
            s = parseInt(secPart);
        }
        return (h * 3600 + m * 60 + s) * 1000 + ms;
    }

    msToTime(ms) {
        let sign = '';
        if (ms < 0) { sign = '-'; ms = -ms; }
        const h = Math.floor(ms / 3600000);
        ms %= 3600000;
        const m = Math.floor(ms / 60000);
        ms %= 60000;
        const s = Math.floor(ms / 1000);
        ms %= 1000;
        return `${sign}${h}:${String(m).padStart(2,'0')}:${String(s).padStart(2,'0')}.${String(ms).padStart(3,'0')}`;
    }

    applyShift(deltaMs) {
        for (const ev of this.events) {
            ev.startMs = Math.max(0, ev.startMs + deltaMs);
            ev.endMs = Math.max(0, ev.endMs + deltaMs);
        }
    }

    applyReplace(oldStr, newStr) {
        for (const ev of this.events) {
            ev.text = ev.text.replace(new RegExp(oldStr, 'g'), newStr);
        }
    }

    applyStyle(newStyle) {
        for (const ev of this.events) {
            ev.style = newStyle;
        }
    }

    save(filename) {
        const lines = [];
        // Сохраняем заголовок
        for (const h of this.header) {
            lines.push(h);
        }
        // Сохраняем события
        for (const ev of this.events) {
            const start = this.msToTime(ev.startMs);
            const end = this.msToTime(ev.endMs);
            // Простая сборка, теряем некоторые поля
            const newLine = `${ev.prefix},${start},${end},${ev.style},0,0,0,,${ev.text}`;
            lines.push(newLine);
        }
        fs.writeFileSync(filename, lines.join('\n'), 'utf8');
        console.log(`Saved to ${filename}`);
    }

    listEvents(color = true) {
        const GREEN = color ? '\x1b[92m' : '';
        const YELLOW = color ? '\x1b[93m' : '';
        const RESET = color ? '\x1b[0m' : '';
        for (let i = 0; i < this.events.length; i++) {
            const ev = this.events[i];
            const start = this.msToTime(ev.startMs);
            const end = this.msToTime(ev.endMs);
            console.log(`${GREEN}[${i}]${RESET} ${start} --> ${end}  ${YELLOW}${ev.style}${RESET}: ${ev.text}`);
        }
    }

    exportSrt(filename) {
        let content = '';
        for (let i = 0; i < this.events.length; i++) {
            const ev = this.events[i];
            const start = this.msToTime(ev.startMs).replace('.', ',');
            const end = this.msToTime(ev.endMs).replace('.', ',');
            content += `${i+1}\n${start} --> ${end}\n${ev.text}\n\n`;
        }
        fs.writeFileSync(filename, content, 'utf8');
        console.log(`Exported SRT to ${filename}`);
    }
}

function main() {
    const args = process.argv.slice(2);
    const opts = {};
    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--input': opts.input = args[++i]; break;
            case '--output': opts.output = args[++i]; break;
            case '--shift': opts.shift = parseInt(args[++i]); break;
            case '--replace': opts.replace = [args[++i], args[++i]]; break;
            case '--style': opts.style = args[++i]; break;
            case '--list': opts.list = true; break;
            case '--export-srt': opts.exportSrt = args[++i]; break;
        }
    }
    if (!opts.input) {
        console.error('Error: --input required');
        process.exit(1);
    }
    const ass = new ASSParser(opts.input);
    ass.parse();

    if (opts.shift !== undefined) ass.applyShift(opts.shift);
    if (opts.replace) ass.applyReplace(opts.replace[0], opts.replace[1]);
    if (opts.style) ass.applyStyle(opts.style);

    if (opts.list) ass.listEvents();
    if (opts.exportSrt) ass.exportSrt(opts.exportSrt);

    if (opts.shift !== undefined || opts.replace || opts.style || opts.output) {
        const out = opts.output || opts.input;
        ass.save(out);
    }
}

if (require.main === module) main();

File: ass_editor.rs
// ass_editor.rs
use std::fs;
use std::io::{self, Write};
use std::path::Path;
use regex::Regex;
use clap::{App, Arg};

struct Event {
    prefix: String,
    start_ms: i32,
    end_ms: i32,
    style: String,
    text: String,
    raw: String,
}

struct ASSParser {
    header: Vec<String>,
    events: Vec<Event>,
}

impl ASSParser {
    fn new() -> Self {
        ASSParser { header: Vec::new(), events: Vec::new() }
    }

    fn parse(&mut self, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
        let content = fs::read_to_string(filename)?;
        let lines: Vec<&str> = content.lines().collect();
        let mut in_events = false;
        for line in lines {
            if line.starts_with("[Events]") {
                in_events = true;
                self.header.push(line.to_string());
                continue;
            }
            if !in_events {
                self.header.push(line.to_string());
                continue;
            }
            if line.starts_with("Format:") {
                self.header.push(line.to_string());
                continue;
            }
            if line.starts_with("Dialogue:") || line.starts_with("Comment:") {
                if let Ok(ev) = Self::parse_event(line) {
                    self.events.push(ev);
                }
            } else {
                self.header.push(line.to_string());
            }
        }
        Ok(())
    }

    fn parse_event(line: &str) -> Result<Event, Box<dyn std::error::Error>> {
        let parts: Vec<&str> = line.splitn(10, ',').collect();
        if parts.len() < 10 {
            return Err("Invalid event format".into());
        }
        let prefix = parts[0].to_string();
        let start_str = parts[1].trim();
        let end_str = parts[2].trim();
        let style = parts[3].trim().to_string();
        let text = parts[9].trim().to_string();
        let start_ms = Self::time_to_ms(start_str)?;
        let end_ms = Self::time_to_ms(end_str)?;
        Ok(Event { prefix, start_ms, end_ms, style, text, raw: line.to_string() })
    }

    fn time_to_ms(ts: &str) -> Result<i32, Box<dyn std::error::Error>> {
        if ts.is_empty() { return Ok(0); }
        let parts: Vec<&str> = ts.split(':').collect();
        if parts.len() != 3 { return Err("Invalid time".into()); }
        let h: i32 = parts[0].parse()?;
        let m: i32 = parts[1].parse()?;
        let sec_part = parts[2];
        let (s, ms) = if let Some((sec, millis)) = sec_part.split_once('.') {
            let s: i32 = sec.parse()?;
            let ms_str = format!("{:0<3}", &millis[..millis.len().min(3)]);
            let ms: i32 = ms_str.parse()?;
            (s, ms)
        } else {
            (sec_part.parse()?, 0)
        };
        Ok((h * 3600 + m * 60 + s) * 1000 + ms)
    }

    fn ms_to_time(ms: i32) -> String {
        let mut sign = "";
        let mut ms_abs = ms;
        if ms < 0 { sign = "-"; ms_abs = -ms; }
        let h = ms_abs / 3600000;
        let rem = ms_abs % 3600000;
        let m = rem / 60000;
        let rem = rem % 60000;
        let s = rem / 1000;
        let ms = rem % 1000;
        format!("{}{}:{:02}:{:02}.{:03}", sign, h, m, s, ms)
    }

    fn apply_shift(&mut self, delta_ms: i32) {
        for ev in &mut self.events {
            ev.start_ms = (ev.start_ms + delta_ms).max(0);
            ev.end_ms = (ev.end_ms + delta_ms).max(0);
        }
    }

    fn apply_replace(&mut self, old: &str, new: &str) {
        let re = Regex::new(old).unwrap();
        for ev in &mut self.events {
            ev.text = re.replace_all(&ev.text, new).to_string();
        }
    }

    fn apply_style(&mut self, new_style: &str) {
        for ev in &mut self.events {
            ev.style = new_style.to_string();
        }
    }

    fn save(&self, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut out = String::new();
        for line in &self.header {
            out.push_str(line);
            out.push('\n');
        }
        for ev in &self.events {
            let start = Self::ms_to_time(ev.start_ms);
            let end = Self::ms_to_time(ev.end_ms);
            let line = format!("{},{},{},{},0,0,0,,{}\n", ev.prefix, start, end, ev.style, ev.text);
            out.push_str(&line);
        }
        fs::write(filename, out)?;
        println!("Saved to {}", filename);
        Ok(())
    }

    fn list_events(&self, color: bool) {
        let (green, yellow, reset) = if color {
            ("\x1b[92m", "\x1b[93m", "\x1b[0m")
        } else {
            ("", "", "")
        };
        for (i, ev) in self.events.iter().enumerate() {
            let start = Self::ms_to_time(ev.start_ms);
            let end = Self::ms_to_time(ev.end_ms);
            println!("{}[{}]{} {} --> {}  {}{}{}: {}", green, i, reset, start, end, yellow, ev.style, reset, ev.text);
        }
    }

    fn export_srt(&self, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
        let mut out = String::new();
        for (i, ev) in self.events.iter().enumerate() {
            let start = Self::ms_to_time(ev.start_ms).replace('.', ',');
            let end = Self::ms_to_time(ev.end_ms).replace('.', ',');
            out.push_str(&format!("{}\n{} --> {}\n{}\n\n", i+1, start, end, ev.text));
        }
        fs::write(filename, out)?;
        println!("Exported SRT to {}", filename);
        Ok(())
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let matches = App::new("ASS Editor")
        .arg(Arg::with_name("input").long("input").takes_value(true).required(true))
        .arg(Arg::with_name("output").long("output").takes_value(true))
        .arg(Arg::with_name("shift").long("shift").takes_value(true))
        .arg(Arg::with_name("replace").long("replace").takes_value(true))
        .arg(Arg::with_name("replace-to").long("replace-to").takes_value(true))
        .arg(Arg::with_name("style").long("style").takes_value(true))
        .arg(Arg::with_name("list").long("list"))
        .arg(Arg::with_name("export-srt").long("export-srt").takes_value(true))
        .get_matches();

    let input = matches.value_of("input").unwrap();
    let mut parser = ASSParser::new();
    parser.parse(input)?;

    if let Some(shift_str) = matches.value_of("shift") {
        let delta: i32 = shift_str.parse()?;
        parser.apply_shift(delta);
    }

    if let (Some(old), Some(new)) = (matches.value_of("replace"), matches.value_of("replace-to")) {
        parser.apply_replace(old, new);
    }

    if let Some(style) = matches.value_of("style") {
        parser.apply_style(style);
    }

    if matches.is_present("list") {
        parser.list_events(true);
    }

    if let Some(srt) = matches.value_of("export-srt") {
        parser.export_srt(srt)?;
    }

    if let Some(output) = matches.value_of("output") {
        parser.save(output)?;
    } else if matches.is_present("shift") || matches.is_present("replace") || matches.is_present("style") {
        parser.save(input)?;
    }

    Ok(())
}
